### PR TITLE
MAINT: Stop vendoring data

### PR DESCRIPTION
**Pull request by thermokarst**

### CODE REVIEW

**Comment by sjanssen2** [comment]
ups, accidentally pressed a button

**Comment by thermokarst** [comment]
Please do not merge this, this is a work in progress.

The recipe is still broken, I am going to refactor a bit to stop vendoring data (this has long been on our todo list, it has represented a big time sink for our build system).

**Comment by thermokarst** [comment]
This PR is now ready for review.

**Comment by sjanssen2** [comment]
Hi @thermokarst can you give a little background about the incentives of this PR? I have the feeling that we loose a lot of functionality. Why is it necessary to cut these things out?

**Comment by thermokarst** [comment]
Hi @sjanssen2 - please point out any lost functionality --- if any was lost it was almost certainly on accident.

The goal of this PR is to stop vendoring the GG database with this plugin, for a few reasons:

- users should be aware of _what_ database they are using --- this PR promotes that by making them provide that information as a required input
- our build system is dragging on builds for this plugin, because of all the vendored data
- vendoring doesn't scale well for other databases.

Anyway, if there is "lost functionality", please point it out, that wasn't the intent. Thanks!

**Comment by sjanssen2** [comment]
Hi @thermokarst ,

> provide that information as a required input
this unfortunately results in a lot of overhead computation, as the user need to qza three files, which than have to be unpacked by the plugin and will take several seconds if not minutes. I tried hard to avoid this situation.
Furthermore, there is no mechanism in place that ensures the three files for the reference belong to each other. The rudimentary check in https://github.com/qiime2/q2-fragment-insertion/blob/8621bb2fc7fedc34292b9e96696388b7cf9c8edd/q2_fragment_insertion/_insertion.py#L42 will be gone by this PR :-/ And yes, checking also takes a while since the full tree needs to be parsed as skbio.TreeNode object.

If vendoring is the only problem, you might want to look into PR #63 , but I have the feeling that this PR not only tried to stop sourcing from this other repo, but also changes the user interface dramatically, i.e. forces the user to always specify three additional parameters - not that comfy :-/

**Comment by thermokarst** [comment]
> If vendoring is the only problem, you might want to look into PR #63

Thanks, I have a few questions about that PR, and will follow up in another message (probably on that PR's issue page). My current gameplan is to merge this PR (#66), then integrate the changes proposed in #63, all in time for the 2019.10 release.

> but I have the feeling that this PR not only tried to stop sourcing from this other repo, but also changes the user interface dramatically, i.e. forces the user to always specify three additional parameters - not that comfy :-/

I assure you, my intent is not to increase user discomfort. Can you comment on the alignment subset size and the placement subset size parameters? Based on the [comments in the code](https://github.com/qiime2/q2-fragment-insertion/pull/66/files#diff-28c8e241dbef6ace3f2223815356e99bL140-L151), my understanding is that the default values are _specific to GG_. If that is the case, then doesn't it make sense to _require users specify that value_ themselves, that way they use the values relevant to the DB in question? I am curious on your thoughts regarding that - thanks.

**Comment by thermokarst** [comment]
Hey @sjanssen2, here is an update on the plan:

- Close #63, but merge the proposed changes there into this PR (#66)
- Retool your new `RAxMLinfoFormat` to be some new "database" format - name pending. The goal here is to address your concern regarding a potential mismatch between database files. This new directory format will contain:
  - reference tree
  - reference alignment
  - reference raxml index
  - default alignment subset size and placement subset size parameters (depends on what you say in reply to [my question above](https://github.com/qiime2/q2-fragment-insertion/pull/66#issuecomment-539059995))
- Publish new QZAs on docs.qiime2.org for GreenGenes and SILVA, for the new database format defined above.
- The user API for running `sepp` will _require_ that users provide one of these database types. The default size parameters will be pulled from the QZA automatically, so no need to change the parameters (we would change the defaults to `None`, that way the plugin would know to use the format's values).

In the end this will look very similar to the workflow in q2-feature-classifier --- the database files in this plugin would be analogous to the trained classifiers in q2-feature-classifier. In the future we could add a new method to this plugin, or perhaps q2-phylogeny, for generating the raxml info file. Also, all the validation of the database inputs would be handled by the new format, so the cost would be up-front (and would be eaten by us, since most users would just download the databases provided on the docs).

I will continue working on these changes here in this PR - thanks!

**Comment by thermokarst** [comment]
This is just about done. The only two remaining TODO items are:

- [ ] cite HMMER & pplacer
- [ ] fetch and import convenience script

With that said, I think this is ready for review, @ebolyen.

File: README.md [hunk 1]
```diff
@@ -1,156 +1,7 @@
...
-```
-qiime fragment-insertion sepp \
-  --i-representative-sequences rep-seqs.qza \
-  --o-tree insertion-tree.qza \
-  --o-placements insertion-placements.qza
-```
-Output artifacts:
-   - `insertion-tree.qza`: ~[view]()~ | [download](https://github.com/biocore/q2-fragment-insertion/blob/master/Example/insertion-tree.qza?raw=true)
-   - `insertion-placements.qza`: ~[view]()~ | [download](https://github.com/biocore/q2-fragment-insertion/blob/master/Example/insertion-placements.qza?raw=true)
+# q2-fragment-insertion
```
Comment (sjanssen2) [inline README.md]: why did you remove the readme content and where can user now find this information? Many people asked e.g. for runtime estimations and this document was my collection of answers to those common questions.

File: q2_fragment_insertion/_format.py [hunk 2]
```diff
@@ -6,16 +6,86 @@
...
+from q2_types.feature_data import AlignedDNAFASTAFormat
+from q2_types.tree import NewickFormat
 
 
 class PlacementsFormat(model.TextFileFormat):
-    def sniff(self):
-        line = open(str(self)).readline()
+    fields = {'tree', 'placements', 'metadata', 'version', 'fields'}
+
+    def _validate_(self, level):
```
Comment (sjanssen2) [inline q2_fragment_insertion/_format.py]: shall we add a comment about where to find a format description? doi.org/10.1371/journal.pone.0031009

File: q2_fragment_insertion/_format.py [hunk 3]
```diff
@@ -6,16 +6,86 @@
...
+    raxml_info = model.File(r'raxml-info.txt', format=RAxMLinfoFormat)
+
+    def _validate_(self, level):
+        seqs = self.alignment.view(skbio.TabularMSA)
+        tree = self.phylogeny.view(skbio.TreeNode)
+
+        seqs.reassign_index(minter='id')
+        alignment_ids = set(seqs.index)
+        phylogeny_ids = {t.name for t in tree.tips()}
+
```
Comment (sjanssen2) [inline q2_fragment_insertion/_format.py]: in principle, we should add a way to check if the info file matches the phylogeny - but I don't know how :-/

File: q2_fragment_insertion/_format.py [hunk 4]
```diff
@@ -6,16 +6,86 @@
...
+                raise ValidationError('Missing structured content: "%s".'
+                                      % sig)
+
+
+class SeppReferenceFormat(model.DirectoryFormat):
+    alignment = model.File(r'aligned-dna-sequences.fasta',
+                           format=AlignedDNAFASTAFormat)
+    phylogeny = model.File(r'tree.nwk', format=NewickFormat)
+    raxml_info = model.File(r'raxml-info.txt', format=RAxMLinfoFormat)
+
```
Comment (sjanssen2) [inline q2_fragment_insertion/_format.py]: didn't we want to add default values for e.g. alignment_subset_size to the reference qza?

File: q2_fragment_insertion/citations.bib [hunk 5]
```diff
@@ -1,10 +1,12 @@
...
-	author={Janssen, Stefan and McDonald, Daniel and Gonzalez, Antonio and Navas-Molina, Jose A. and Jiang, Lingjing and Xu, Zhenjiang Zech and Winker, Kevin and Kado, Deborah M. and Orwoll, Eric and Manary, Mark and Mirarab, Siavash and Knight, Rob},
+  author={Janssen, Stefan and McDonald, Daniel and Gonzalez, Antonio and Navas-Molina, Jose A. and Jiang, Lingjing and Xu, Zhenjiang Zech and Winker, Kevin and Kado, Deborah M. and Orwoll, Eric and Manary, Mark and Mirarab, Siavash and Knight, Rob},
   journal={mSystems},
   volume={3},
   number={3},
   year={2018},
   publisher={American Society for Microbiology Journals},
   doi={10.1128/mSystems.00021-18},
 }
+
```
Comment (sjanssen2) [inline q2_fragment_insertion/citations.bib]: # from current docu of HMMER: "There has never been a paper on the HMMER software. The best citation is to the web site, hmmer.org."

@article{hmmer,
    author={Eddy, Sean R.},
    journal={PLOS Computational Biology},
    publisher={Public Library of Science},
    title={Accelerated Profile HMM Searches},
    year={2011},
    month={10},
    volume={7},
    url={https://doi.org/10.1371/journal.pcbi.1002195},
    pages={1-16},
    number={10},
    doi={10.1371/journal.pcbi.1002195}
}
@article{pplacer,
  title={pplacer: linear time maximum-likelihood and Bayesian phylogenetic placement of sequences onto a fixed reference tree},
  author={Matsen, Frederick A and Kodner, Robin B and Armbrust, E Virginia},
  journal={BMC bioinformatics},
  volume={11},
  number={1},
  pages={538},
  year={2010},
  publisher={BioMed Central}
}

File: q2_fragment_insertion/_format.py (hunk 4 above)
Comment (thermokarst) [reply q2_fragment_insertion/_format.py]: Maybe? I was waiting to hear back from you on that:

> Based on the comments in the code, my understanding is that the default values are specific to GG. If that is the case, then doesn't it make sense to require users specify that value themselves, that way they use the values relevant to the DB in question? I am curious on your thoughts regarding that - thanks.

My initial thought was "yes", but then I noticed in #63 you didn't make any notes about these values changing for Silva (or if you did, I missed them). Can you comment on this?

File: README.md (hunk 1 above)
Comment (thermokarst) [reply README.md]: This information will be moved to library.qiime2.org

**Comment by thermokarst** [comment]
> Should we somehow add commands on how to create reference sets for GG and/or Silva, such that we can re-produce those in the future after we ourselves forgot how to?

Yes, that is one of my TODOs [listed above](https://github.com/qiime2/q2-fragment-insertion/pull/66#issuecomment-546141311).

File: q2_fragment_insertion/_format.py (hunk 2 above)
Comment (thermokarst) [reply q2_fragment_insertion/_format.py]: Great, thanks for this, will add!

File: q2_fragment_insertion/_format.py (hunk 3 above)
Comment (thermokarst) [reply q2_fragment_insertion/_format.py]: I have some ideas on that, stay tuned.

File: q2_fragment_insertion/citations.bib (hunk 5 above)
Comment (thermokarst) [reply q2_fragment_insertion/citations.bib]: Perfect, thanks for doing the legwork here!

File: q2_fragment_insertion/_format.py (hunk 3 above)
Comment (thermokarst) [reply q2_fragment_insertion/_format.py]: Going to skip this for now, see the new comment, in source.

File: q2_fragment_insertion/_format.py (hunk 4 above)
Comment (sjanssen2) [reply q2_fragment_insertion/_format.py]: for now, I don't have those values for Silva or other references. I hope to find a student volunteer soon to benchmark Silva. Thus, I agree with your assessment and vote against moving those parameters to the reference set

Comment (sjanssen2) [reply q2_fragment_insertion/_format.py]: that said: is there a way to overwrite regular parameters for a qiime2 function, i.e. can we optionally provide better suiting default per reference in the future? Thus, the normal user would not need to worry about both parameters

**Comment by Rkubinski** [comment]
Hi Guys, so to my understanding, we will soon be able to pass any reference alignment into q2-fragment-insertion? Is that correct ?

**Comment by thermokarst** [comment]
Hi @Rkubinski!

> Is that correct ?

Close, but not quite. q2-fragment-insertion 2019.10 will support both of the references shipped in sepp-refs: https://github.com/smirarab/sepp-refs/. As I mentioned in my email response to you yesterday, we aren't really involved in the preparation or creation of sepp-refs - please contact the sepp-refs team by opening an issue up at https://github.com/smirarab/sepp-refs/issues. Thanks!
